File: macos-swiftui/Tests/ArchiveConformanceTests.swift
/* The conformance test docs/archive-layout.md asks every out-of-repo consumer
 * to keep: build a fixture tree in the documented shape and assert this reader
 * finds it.
 *
 * The point is not coverage. It is that a layout bump in the pipeline becomes a
 * FAILING BUILD here rather than a user reporting an empty library. This app's
 * tests passing and the pipeline's tests passing says nothing about the two
 * agreeing; only a fixture in the documented shape does.
 *
 * Every case below is a state the contract names as ordinary and that a naive
 * reader gets wrong -- particularly the layout 1 -> 2 changes, where matching on
 * ".mkv" was correct and is now a bug. The thirteen the GTK suite carries are
 * the floor, not the target; the last two here are macOS's own.
 */

import XCTest
@testable import YtdlMac

final class ArchiveConformanceTests: XCTestCase {
    var tmpdir: String = ""
    /// The fake "Complete Archive".
    var root: String = ""

    override func setUpWithError() throws {
        tmpdir = try FixtureSupport.makeTempDir(prefix: "ytdl-macos-archive")
        root = Paths.join(tmpdir, "Youtube Videos/Complete Archive")
        try FixtureSupport.mkdirp(root)
    }

    override func tearDownWithError() throws {
        Paths.environmentOverride = nil
        try? FileManager.default.removeItem(atPath: tmpdir)
    }

    // MARK: - Fixture

    /// One video folder in the documented shape. `manifestExtra` is spliced into
    /// manifest.json; `info` is written verbatim as the info.json, or nil to
    /// exercise the folder-name fallback.
    @discardableResult
    private func makeVideo(
        channel: String,
        folder: String,
        manifestExtra: String? = nil,
        info: String? = nil
    ) throws -> String {
        let vdir = Paths.join(Paths.join(root, channel), folder)
        try FixtureSupport.mkdirp(vdir)

        let extra = manifestExtra.map { ",\n  \($0)" } ?? ""
        let manifest = """
        {
          "archive_layout_version": 2,
          "video_id": "dQw4w9WgXcQ",
          "title": "from the manifest",
          "uploader": "\(channel)"\(extra)
        }
        """
        try FixtureSupport.write(
            manifest, to: Paths.join(vdir, "Video metadata/manifest.json")
        )
        if let info {
            try FixtureSupport.write(
                info, to: Paths.join(vdir, "Video metadata/Video.info.json")
            )
        }
        return vdir
    }

    private func scan() throws -> ArchiveIndex {
        try ArchiveIndex.scan(root: root)
    }

    private func onlyEntry(_ index: ArchiveIndex) throws -> ArchiveEntry {
        XCTAssertEqual(index.entries.count, 1)
        return try XCTUnwrap(index.entries.first)
    }

    // MARK: - Cases

    func testDiscoversContractLayout() throws {
        let v = try makeVideo(
            channel: "Rick Astley",
            folder: "Rick Astley - 20091025 - dQw4w9WgXcQ - Never Gonna Give You Up",
            manifestExtra: "\"media_file\": \"Final files/Final Video.mkv\""
        )
        try FixtureSupport.write(
            "not really a video", to: Paths.join(v, "Final files/Final Video.mkv")
        )

        let index = try scan()
        let e = try onlyEntry(index)

        XCTAssertEqual(e.channel, "Rick Astley")
        XCTAssertEqual(e.videoID, "dQw4w9WgXcQ")
        XCTAssertEqual(e.uploadDate, "20091025")
        XCTAssertNotNil(e.mediaIndex)

        // The key round-trips: it is the only handle the UI ever holds.
        XCTAssertEqual(index.entry(forKey: e.key)?.key, e.key)

        XCTAssertEqual(index.videoCount, 1)
        XCTAssertEqual(index.channelCount, 1)
    }

    func testFolderNameFallback() throws {
        // No info.json at all -- the documented fallback, and a real state.
        try makeVideo(
            channel: "Some Channel",
            folder: "Some Channel - 20240101 - abcdefghijk - A Title - With Dashes"
        )

        let e = try onlyEntry(try scan())
        XCTAssertEqual(e.videoID, "abcdefghijk")
        XCTAssertEqual(e.uploadDate, "20240101")
        /* The title keeps its own " - ", which a naive four-field split would
         * truncate to "A Title". */
        XCTAssertEqual(e.title, "A Title - With Dashes")
    }

    func testMediaFoundWhateverContainer() throws {
        for (i, ext) in ["mkv", "mp4", "webm"].enumerated() {
            /* No media_file in the manifest: this is the GLOB path, which is the
             * one that was wrong under layout 1. */
            let v = try makeVideo(
                channel: "Chan",
                folder: "Chan - 2024010\(i + 1) - abcdefghij\(i) - T"
            )
            try FixtureSupport.write("x", to: Paths.join(v, "Final files/Final Video.\(ext)"))
        }

        let index = try scan()
        XCTAssertEqual(index.entries.count, 3)
        for e in index.entries {
            XCTAssertNotNil(e.mediaIndex, "no media found for \(e.rel)")
        }
    }

    func testAudioOnlyDownload() throws {
        /* Layout 2's headline change: Final AUDIO, and an extension a
         * video-only reader would never glob for. */
        let v = try makeVideo(
            channel: "Chan",
            folder: "Chan - 20240101 - abcdefghijk - T",
            manifestExtra: "\"download_mode\": \"audio-only\""
        )
        try FixtureSupport.write("x", to: Paths.join(v, "Final files/Final Audio.opus"))

        let e = try onlyEntry(try scan())
        let mi = try XCTUnwrap(e.mediaIndex)
        XCTAssertEqual(e.files[mi].ext, ".opus")
        XCTAssertEqual(e.downloadMode, "audio-only")
    }

    func testNoMediaIsValid() throws {
        /* --mode metadata-only writes a complete folder with no media in it.
         * The entry must EXIST and report nil, not be hidden and not be an
         * error. */
        try makeVideo(
            channel: "Chan",
            folder: "Chan - 20240101 - abcdefghijk - T",
            manifestExtra: "\"download_mode\": \"metadata-only\", \"media_file\": null"
        )

        let e = try onlyEntry(try scan())
        XCTAssertNil(e.mediaIndex)
        XCTAssertEqual(e.downloadMode, "metadata-only")
    }

    func testSkipsPreMergeStreams() throws {
        /* --keep-video leaves video-only and audio-only files behind. Picking
         * one gives a silent video or a black audio track, so the folder is
         * skipped outright -- and the format-id name is refused even in Final
         * files/. */
        let v = try makeVideo(channel: "Chan", folder: "Chan - 20240101 - abcdefghijk - T")
        try FixtureSupport.write(
            "silent video", to: Paths.join(v, "Pre-merge streams/Final Video.f137.mp4")
        )
        try FixtureSupport.write(
            "audio only", to: Paths.join(v, "Final files/Final Video.f251.webm")
        )

        XCTAssertNil(try onlyEntry(try scan()).mediaIndex)

        // Add the real merged output and it wins.
        try FixtureSupport.write(
            "the actual video", to: Paths.join(v, "Final files/Final Video.mkv")
        )
        let e = try onlyEntry(try scan())
        let mi = try XCTUnwrap(e.mediaIndex)
        XCTAssertEqual(e.files[mi].rel, "Final files/Final Video.mkv")
    }

    func testVideoWinsOverAudio() throws {
        let v = try makeVideo(channel: "Chan", folder: "Chan - 20240101 - abcdefghijk - T")
        /* Written in an order where a directory listing could hand back either
         * first; the choice must not depend on that. */
        try FixtureSupport.write("a", to: Paths.join(v, "Final files/Final Audio.m4a"))
        try FixtureSupport.write("v", to: Paths.join(v, "Final files/Final Video.mp4"))

        let e = try onlyEntry(try scan())
        let mi = try XCTUnwrap(e.mediaIndex)
        XCTAssertEqual(e.files[mi].ext, ".mp4")
    }

    func testManifestMediaFileWins() throws {
        /* The contract says prefer media_file over globbing. Here globbing would
         * find the .mkv; the manifest names the .mp4. */
        let v = try makeVideo(
            channel: "Chan",
            folder: "Chan - 20240101 - abcdefghijk - T",
            manifestExtra: "\"media_file\": \"Final files/Final Video.mp4\""
        )
        try FixtureSupport.write("x", to: Paths.join(v, "Final files/Final Video.mkv"))
        try FixtureSupport.write("y", to: Paths.join(v, "Final files/Final Video.mp4"))

        let e = try onlyEntry(try scan())
        let mi = try XCTUnwrap(e.mediaIndex)
        XCTAssertEqual(e.files[mi].rel, "Final files/Final Video.mp4")
    }

    func testMissingLayoutVersionIsLayoutOne() throws {
        /* An absent field means the video predates versioning, which the
         * contract defines as layout 1 -- readable, not an error. */
        let vdir = Paths.join(root, "Chan/Chan - 20240101 - abcdefghijk - T")
        try FixtureSupport.write(
            "{ \"video_id\": \"abcdefghijk\" }",
            to: Paths.join(vdir, "Video metadata/manifest.json")
        )
        try FixtureSupport.write("x", to: Paths.join(vdir, "Final files/Final Video.mkv"))

        let e = try onlyEntry(try scan())
        XCTAssertEqual(e.layoutVersion, 0)
        XCTAssertFalse(e.layoutTooNew)
        XCTAssertNotNil(e.mediaIndex)
    }

    func testFlagsNewerLayout() throws {
        let vdir = Paths.join(root, "Chan/Chan - 20240101 - abcdefghijk - T")
        try FixtureSupport.write(
            "{ \"archive_layout_version\": \(supportedArchiveLayout + 1) }",
            to: Paths.join(vdir, "Video metadata/manifest.json")
        )

        /* Flagged, but still PRESENT. An empty library with no explanation is
         * the outcome the whole contract exists to prevent. */
        XCTAssertTrue(try onlyEntry(try scan()).layoutTooNew)
    }

    func testChannelInfoIsNotAVideo() throws {
        try makeVideo(channel: "Chan", folder: "Chan - 20240101 - abcdefghijk - T")
        try FixtureSupport.write(
            "not a video", to: Paths.join(root, "Chan/Channel Info/avatar.png")
        )

        XCTAssertEqual(try scan().entries.count, 1)
    }

    func testPathForIndexStaysInside() throws {
        let v = try makeVideo(channel: "Chan", folder: "Chan - 20240101 - abcdefghijk - T")
        try FixtureSupport.write("x", to: Paths.join(v, "Final files/Final Video.mkv"))

        let e = try onlyEntry(try scan())
        let mi = try XCTUnwrap(e.mediaIndex)
        let path = try XCTUnwrap(e.path(forIndex: mi))
        XCTAssertTrue(path.hasPrefix(e.dir))
        XCTAssertTrue(Paths.isRegularFile(path))

        // Out of range returns nil rather than reading past the array.
        XCTAssertNil(e.path(forIndex: e.files.count + 99))
        XCTAssertNil(e.path(forIndex: -1))
    }

    func testScanReportsAMissingRoot() {
        XCTAssertThrowsError(
            try ArchiveIndex.scan(root: Paths.join(tmpdir, "no-such-archive"))
        )
    }

    /* Not a fixture test: pure parsing, and the case that motivated anchoring
     * the parse on the date+id shape rather than splitting into four fields. */
    func testFolderNameParsing() throws {
        let parsed = try XCTUnwrap(FolderName.parse("A - B - 20240101 - abcdefghijk - T - U"))
        XCTAssertEqual(parsed.uploader, "A - B")
        XCTAssertEqual(parsed.uploadDate, "20240101")
        XCTAssertEqual(parsed.id, "abcdefghijk")
        XCTAssertEqual(parsed.title, "T - U")

        /* Nothing resembling a date and an id: the caller falls back to the
         * whole folder name, so this must say so rather than half-fill. */
        XCTAssertNil(FolderName.parse("just a folder"))
    }

    /* CLI_VERSION's REQUIRES_ARCHIVE_LAYOUT and Archive.swift's
     * supportedArchiveLayout are two copies of one fact. The pin is what an
     * installer and a human read; the constant is what actually decides, per
     * video, whether a folder is rendered or flagged.
     *
     * ONE NUMBER, ALL THREE APPS -- CLI_VERSION belongs to the repository, not
     * to this app, and the GTK suite asserts against the same line. */
    func testPinMatchesSupportedLayout() throws {
        let path = try XCTUnwrap(
            FixtureSupport.repositoryFile(named: "CLI_VERSION"),
            "CLI_VERSION not found; the test target's resource reference is missing"
        )
        let pin = try XCTUnwrap(CLIVersion.load(atPath: path))
        let declared = try XCTUnwrap(pin.requiresArchiveLayout)

        XCTAssertGreaterThan(declared, 0)
        XCTAssertEqual(declared, supportedArchiveLayout)
    }

    // MARK: - macOS's own two

    /* The key is not local to this app: it is a hash of the archive-relative
     * folder path, and the GTK app computes the same one for the same folder.
     * Pinned to a literal so a "harmless" change to the hashing -- a different
     * truncation, a normalisation, hashing the absolute path -- fails here
     * rather than silently giving this app its own private addressing. */
    func testKeyMatchesTheOtherApps() {
        XCTAssertEqual(
            Paths.key(for: "Rick Astley/Rick Astley - 20091025 - dQw4w9WgXcQ - Never Gonna Give You Up"),
            "e3732acf5aa3ba15"
        )
        XCTAssertEqual(Paths.key(for: "Chan/Chan - 20240101 - abcdefghijk - T"),
                       "e75480efc59e3e13")
        // Windows-written manifests spell the separator the other way.
        XCTAssertEqual(Paths.key(for: "Chan\\Chan - 20240101 - abcdefghijk - T"),
                       Paths.key(for: "Chan/Chan - 20240101 - abcdefghijk - T"))
    }

    /* Derived state goes to this app's OWN directories under ~/Library. A
     * shared cache would mean three readers with three index formats each
     * treating the others' files as corrupt, and a state directory under
     * Caches would mean macOS deleting the user's profiles when the disk
     * fills. */
    func testCacheAndStateAreSeparateAndOurs() {
        Paths.environmentOverride = ["HOME": tmpdir]
        defer { Paths.environmentOverride = nil }

        XCTAssertEqual(Paths.cacheDir(), tmpdir + "/Library/Caches/ytdl-macos")
        XCTAssertEqual(Paths.stateDir(), tmpdir + "/Library/Application Support/ytdl-macos")
        XCTAssertNotEqual(Paths.cacheDir(), Paths.stateDir())
        XCTAssertFalse(Paths.stateDir().contains("/Caches/"))
        XCTAssertFalse(Paths.cacheDir().contains("ytdl-gtk"))
    }

    /* The acceptance set for --archive-root, which must match
     * archive-viewer.py's: a path that works for one works for both. */
    func testArchiveRootAcceptanceSet() throws {
        try makeVideo(channel: "Chan", folder: "Chan - 20240101 - abcdefghijk - T")

        let dataRoot = tmpdir
        let youtubeVideos = Paths.join(tmpdir, "Youtube Videos")
        let channelDir = Paths.join(root, "Chan")

        XCTAssertEqual(Paths.resolveArchiveRoot(dataRoot), root)
        XCTAssertEqual(Paths.resolveArchiveRoot(youtubeVideos), root)
        XCTAssertEqual(Paths.resolveArchiveRoot(root), root)
        XCTAssertEqual(Paths.resolveArchiveRoot(channelDir), root)
        XCTAssertNil(Paths.resolveArchiveRoot(Paths.join(tmpdir, "nowhere")))
    }
}

File: windows-winui/YtdlWin.Tests/ArchiveConformanceTests.cs
/* The archive conformance suite.
 *
 * docs/archive-layout.md in the pipeline repo says what this is for:
 *
 *   "Out-of-repo consumers are expected to keep their own conformance test that
 *    builds a fixture tree in this shape and asserts their reader finds it.
 *    That test is what turns a bump into a build failure on their side rather
 *    than a bug report from a user."
 *
 * So this file is not optional and an app in this repository without one does
 * not belong in it. It covers the cases a LAYOUT-1 reader gets wrong, because
 * those are the ones that fail silently: a folder that indexes fine and shows
 * the wrong file, or none.
 *
 * It also covers the two failures that are this platform's alone -- separators
 * and MAX_PATH -- which the GTK and SwiftUI suites have no reason to test and
 * which would each produce a short library with no error anywhere.
 */

using System;
using System.IO;
using System.Linq;
using Xunit;
using YtdlWin.Core;

namespace YtdlWin.Tests;

public sealed class ArchiveConformanceTests : IDisposable
{
    private readonly string _root;
    private readonly string _archive;

    public ArchiveConformanceTests()
    {
        _root = FixtureSupport.MakeTempDir("ytdl-win-archive");
        _archive = Path.Combine(_root, "Youtube Videos", "Complete Archive");
        Directory.CreateDirectory(_archive);
    }

    public void Dispose() => FixtureSupport.Delete(_root);

    // MARK: - Fixture building

    private string MakeVideo(string channel, string folder)
    {
        var dir = Path.Combine(_archive, channel, folder);
        Directory.CreateDirectory(Path.Combine(dir, "Final files"));
        Directory.CreateDirectory(Path.Combine(dir, "Video metadata"));
        return dir;
    }

    private static void WriteManifest(string videoDir, string body)
        => FixtureSupport.Write(body, Path.Combine(videoDir, "Video metadata", "manifest.json"));

    private static void WriteFile(string videoDir, string relative, string contents = "x")
        => FixtureSupport.Write(contents, Path.Combine(videoDir, relative.Replace('/', Path.DirectorySeparatorChar)));

    private const string Folder = "Rick Astley - 20091025 - dQw4w9WgXcQ - Never Gonna Give You Up";

    private ArchiveEntry ScanOne()
    {
        var index = ArchiveIndex.Scan(_archive);
        return Assert.Single(index.Entries);
    }

    // MARK: - The pin

    /* Two copies of one fact, asserted to agree. The pin is what an installer
     * and a human read; the constant is what actually decides, per video,
     * whether a folder is rendered or flagged. Them disagreeing is how an app
     * ends up claiming to read a layout it does not.
     *
     * Reads the FILE rather than hardcoding the number, which is the whole
     * point -- a test with the number written into it stops being a test of
     * agreement the moment somebody edits one side. */
    [Fact]
    public void SupportedLayoutMatchesTheRepositoryPin()
    {
        var path = FixtureSupport.RepositoryFile("CLI_VERSION");
        Assert.True(path is not null,
            "CLI_VERSION was not found beside the test binary or above it. It is copied by the " +
            "test project; a missing copy means this assertion is not running, not that it passed.");

        var pin = CliVersion.Load(path!);
        Assert.NotNull(pin);
        Assert.True(pin!.RequiresArchiveLayout.HasValue,
            "CLI_VERSION has no REQUIRES_ARCHIVE_LAYOUT line.");
        Assert.Equal(ArchiveLayout.Supported, pin.RequiresArchiveLayout!.Value);
    }

    // MARK: - Layout 2 media rules

    /* THE LAYOUT-1 BUG, stated once. A reader that globs for *.mkv was correct
     * under layout 1 and is wrong under layout 2, where --mode audio-only writes
     * Final Audio.opus and --container writes .mp4 or .webm. Matching on the
     * BASE NAME is the rule. */
    [Theory]
    [InlineData("Final files/Final Video.mkv")]
    [InlineData("Final files/Final Video.mp4")]
    [InlineData("Final files/Final Video.webm")]
    [InlineData("Final files/Final Audio.opus")]
    [InlineData("Final files/Final Audio.m4a")]
    public void MediaIsFoundByBaseNameWhateverTheExtension(string relative)
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, relative);

        var entry = ScanOne();
        Assert.True(entry.MediaIndex >= 0, $"{relative} was not recognised as the media file");
        Assert.Equal(relative, entry.Files[entry.MediaIndex].Rel);
    }

    /* The manifest's media_file wins over the glob, because it is the only
     * answer that stays right when --container changes the extension. */
    [Fact]
    public void ManifestMediaFileWinsOverTheGlob()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteFile(dir, "Final files/Final Audio.opus");
        WriteManifest(dir, """
        {
          "archive_layout_version": 2,
          "media_file": "Final files/Final Audio.opus"
        }
        """);

        var entry = ScanOne();
        Assert.Equal("Final files/Final Audio.opus", entry.Files[entry.MediaIndex].Rel);
    }

    /* WINDOWS ONLY, and the reason this suite is not just a translation of the
     * Swift one. postprocess.ps1 runs on Windows here, and a manifest written
     * on this platform spells media_file with BACKSLASHES. Compared against a
     * '/'-separated rel that matches nothing -- and the reader then falls
     * through to the glob, which usually gets the right answer anyway. That is
     * exactly what makes it the kind of bug that survives for a year: it is
     * only visible when the glob and the manifest disagree, which is what this
     * fixture arranges. */
    [Fact]
    public void ManifestMediaFileWithBackslashesStillMatches()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteFile(dir, "Final files/Final Audio.opus");
        WriteManifest(dir, """
        {
          "archive_layout_version": 2,
          "media_file": "Final files\\Final Audio.opus"
        }
        """);

        var entry = ScanOne();
        Assert.Equal("Final files/Final Audio.opus", entry.Files[entry.MediaIndex].Rel);
    }

    /* A manifest naming a file somebody has since deleted degrades to the glob,
     * not to a broken path. */
    [Fact]
    public void ManifestNamingAMissingFileFallsBackToTheGlob()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteManifest(dir, """
        { "archive_layout_version": 2, "media_file": "Final files/Final Audio.opus" }
        """);

        var entry = ScanOne();
        Assert.Equal("Final files/Final Video.mkv", entry.Files[entry.MediaIndex].Rel);
    }

    /* A folder with NO media at all is VALID. --mode metadata-only,
     * comments-only and subs-only all write one, and so does an interrupted
     * run. A reader that treats this as corruption hides real archive
     * content. */
    [Fact]
    public void AFolderWithNoMediaIsOrdinaryAndStillIndexed()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Video metadata/Info.info.json", "{}");
        WriteManifest(dir, """
        { "archive_layout_version": 2, "download_mode": "metadata-only" }
        """);

        var entry = ScanOne();
        Assert.Equal(-1, entry.MediaIndex);
        Assert.Equal("metadata-only", entry.DownloadMode);
        Assert.Equal("Never Gonna Give You Up", entry.Title);
    }

    /* Pre-merge streams/ holds --keep-video's video-only and audio-only files.
     * Picking one gives a silent video or a black audio track. */
    [Fact]
    public void PreMergeStreamsAreNeverChosen()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Pre-merge streams/Final Video.mkv");
        WriteFile(dir, "Final files/Final Video.mkv");

        var entry = ScanOne();
        Assert.Equal("Final files/Final Video.mkv", entry.Files[entry.MediaIndex].Rel);
    }

    [Fact]
    public void PreMergeStreamsAloneMeansNoMedia()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Pre-merge streams/Final Video.mkv");

        Assert.Equal(-1, ScanOne().MediaIndex);
    }

    /* A format-id segment names a --keep-video leftover, and is excluded even
     * inside Final files/. */
    [Theory]
    [InlineData("Final Video.f137.mp4", true)]
    [InlineData("Final Video.f251.webm", true)]
    [InlineData("Final Video.mkv", false)]
    [InlineData("Final Video.f.mp4", false)]      // "f" with no digits
    [InlineData("Final Video.fx37.mp4", false)]   // not all digits
    [InlineData("Final Video.2023.mp4", false)]   // digits, but no leading f
    public void FormatIdSegmentsAreRecognised(string name, bool expected)
        => Assert.Equal(expected, ArchiveEntry.HasFormatIdSegment("Final files/" + name));

    [Fact]
    public void FormatIdFilesAreNotChosenAsTheMedia()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.f137.mp4");

        Assert.Equal(-1, ScanOne().MediaIndex);
    }

    [Fact]
    public void VideoWinsOverAudioWhenAFolderHoldsBoth()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Audio.opus");
        WriteFile(dir, "Final files/Final Video.mkv");

        // One scan, not two: indexing into the files of a second scan with an
        // index taken from the first is a bug waiting for the day the two
        // disagree.
        var entry = ScanOne();
        Assert.Equal("Final files/Final Video.mkv", entry.Files[entry.MediaIndex].Rel);
    }

    // MARK: - Layout version

    [Fact]
    public void AManifestWithNoVersionIsLayoutOneAndNotAnError()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteManifest(dir, """{ "video_id": "dQw4w9WgXcQ" }""");

        var entry = ScanOne();
        Assert.Equal(0, entry.LayoutVersion);
        Assert.False(entry.LayoutTooNew);
    }

    /* A video written with a NEWER layout is shown and FLAGGED, never hidden.
     * An empty library with no explanation is the outcome the contract exists
     * to prevent. */
    [Fact]
    public void ANewerLayoutIsFlaggedButStillIndexed()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteManifest(dir, $$"""
        { "archive_layout_version": {{ArchiveLayout.Supported + 1}} }
        """);

        var entry = ScanOne();
        Assert.True(entry.LayoutTooNew);
        Assert.True(entry.MediaIndex >= 0);
    }

    // MARK: - The folder-name fallback

    /* The documented fallback when info.json is missing or unparseable. Anchors
     * on the date and the id because BOTH the uploader and the title routinely
     * contain " - " themselves. */
    [Fact]
    public void FolderNameIsParsedWhenThereIsNoMetadata()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");

        var entry = ScanOne();
        Assert.Equal("Rick Astley", entry.Uploader);
        Assert.Equal("20091025", entry.UploadDate);
        Assert.Equal("dQw4w9WgXcQ", entry.VideoId);
        Assert.Equal("Never Gonna Give You Up", entry.Title);
    }

    [Fact]
    public void SeparatorsInsideTheUploaderAndTitleSurvive()
    {
        var parsed = FolderName.Parse("Some - Band - 20240131 - abcdefghijk - A - B - C");
        Assert.NotNull(parsed);
        Assert.Equal("Some - Band", parsed!.Uploader);
        Assert.Equal("20240131", parsed.UploadDate);
        Assert.Equal("abcdefghijk", parsed.Id);
        Assert.Equal("A - B - C", parsed.Title);
    }

    [Theory]
    [InlineData("no separators at all")]
    [InlineData("Uploader - notadate - abcdefghijk - Title")]
    [InlineData("Uploader - 20240131 - tooshort - Title")]
    public void AnUnparseableFolderNameYieldsNull(string name)
        => Assert.Null(FolderName.Parse(name));

    [Fact]
    public void AnUnparseableFolderNameFallsBackToTheWholeName()
    {
        const string odd = "some folder nobody named properly";
        var dir = MakeVideo("Rick Astley", odd);
        WriteFile(dir, "Final files/Final Video.mkv");

        var entry = ScanOne();
        Assert.Equal(odd, entry.Title);
        Assert.Equal("Rick Astley", entry.Uploader);   // falls back to the channel folder
        Assert.Null(entry.VideoId);
    }

    // MARK: - Structure

    [Fact]
    public void ChannelInfoIsNotAVideo()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        Directory.CreateDirectory(Path.Combine(_archive, "Rick Astley", "Channel Info"));
        FixtureSupport.Write("{}", Path.Combine(_archive, "Rick Astley", "Channel Info", "avatar.json"));

        var index = ArchiveIndex.Scan(_archive);
        Assert.Single(index.Entries);
        Assert.Single(index.Channels);
    }

    [Fact]
    public void AChannelWithNoVideosIsNotCounted()
    {
        Directory.CreateDirectory(Path.Combine(_archive, "Empty Channel"));
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");

        var index = ArchiveIndex.Scan(_archive);
        Assert.Equal(new[] { "Rick Astley" }, index.Channels);
    }

    [Fact]
    public void ScanningANonDirectoryThrowsAndNamesTheRoot()
    {
        var missing = Path.Combine(_root, "not-here");
        var ex = Assert.Throws<ArchiveIndex.ScanException>(() => ArchiveIndex.Scan(missing));
        Assert.Contains(missing, ex.Message, StringComparison.Ordinal);
    }

    // MARK: - The opaque key and path resolution

    /* The key is SHA-256 of the '/'-separated relative path truncated to 8
     * bytes. It must come out byte for byte the same as the key the GTK and
     * SwiftUI apps compute for the same folder -- so this pins the ALGORITHM
     * with a fixed vector rather than merely checking two calls agree, which
     * would pass just as happily if the whole scheme changed. */
    [Fact]
    public void TheKeyIsSixteenHexCharactersOfSha256()
    {
        var key = Paths.Key("Rick Astley/" + Folder);
        Assert.Equal(16, key.Length);
        Assert.All(key, c => Assert.True(char.IsAsciiDigit(c) || (c >= 'a' && c <= 'f'),
            $"'{c}' is not lowercase hex"));

        // Backslashes normalise, so the same folder keys the same whichever
        // platform indexed it. This is the assertion that would fail first if a
        // Path.Combine ever leaked a native separator into a `rel`.
        Assert.Equal(key, Paths.Key(@"Rick Astley\" + Folder));
    }

    [Fact]
    public void PathResolutionRefusesToEscapeTheFolder()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        var entry = ScanOne();

        Assert.Null(entry.PathForIndex(-1));
        Assert.Null(entry.PathForIndex(entry.Files.Count));

        var resolved = entry.PathForIndex(entry.MediaIndex);
        Assert.NotNull(resolved);
        Assert.True(Paths.IsInside(entry.Dir, resolved!));

        // A hand-made entry claiming a traversing relative path is refused --
        // the containment re-check, not a filter, is what makes that true.
        var hostile = new ArchiveEntry
        {
            Key = "", Dir = entry.Dir, Rel = "", Channel = "",
            Files = { new ArchiveFile { Rel = "../../escape.txt", Ext = ".txt", Folder = "", Size = 0 } },
        };
        Assert.Null(hostile.PathForIndex(0));
    }

    /* NTFS is case-insensitive, so a containment check that compared ordinally
     * would reject a legitimate path whenever the two halves were spelled with
     * different capitalisation -- which happens routinely, because one came
     * from a folder picker and the other from a typed setting. */
    [Fact]
    public void ContainmentIsCaseInsensitive()
    {
        Assert.True(Paths.IsInside(@"C:\Archive", @"c:\archive\video\file.mkv"));
        Assert.True(Paths.IsInside(@"C:\Archive", @"C:\Archive"));
        // ...and "C:\Archive" does not contain "C:\Archive2".
        Assert.False(Paths.IsInside(@"C:\Archive", @"C:\Archive2\file.mkv"));
    }

    // MARK: - Thumbnails

    [Fact]
    public void ImagesFolderBeatsALargerImageElsewhere()
    {
        var dir = MakeVideo("Rick Astley", Folder);
        WriteFile(dir, "Final files/Final Video.mkv");
        WriteFile(dir, "Images/thumb.jpg", "small");
        WriteFile(dir, "Final files/cover.png", new string('x', 50_000));

        var entry = ScanOne();
        Assert.Equal("Images/thumb.jpg", entry.Files[entry.ThumbnailIndex].Rel);
    }

    // MARK: - MAX_PATH

    /* THIS PLATFORM'S OWN FAILURE, and it has no counterpart in the GTK or
     * SwiftUI suites.
     *
     * The pipeline's per-video paths run to roughly 240 characters before the
     * data root is prefixed. Past 260, an ordinary Win32 path fails -- and it
     * does not throw somewhere visible: the directory enumeration comes back
     * empty, the deepest files (Video metadata/, Pre-merge streams/) are
     * missing from the listing, and the library shows a video with no metadata
     * and no error anywhere.
     *
     * The fixture builds a genuinely over-long path rather than asserting on
     * Paths.Extended's string output, because the string transformation being
     * right is not the claim -- the claim is that the READER finds the file. */
    [Fact]
    public void DeeplyNestedFilesPastMaxPathAreStillFound()
    {
        // A 90-character uploader and a 120-character title, which is long but
        // well within what YouTube allows and what this pipeline produces.
        var channel = new string('U', 90);
        var folder = $"{channel} - 20091025 - dQw4w9WgXcQ - {new string('T', 120)}";

        var dir = Path.Combine(_archive, channel, folder);
        var deep = Path.Combine(dir, "Video metadata");

        try
        {
            Directory.CreateDirectory(Paths.Extended(deep));
            Directory.CreateDirectory(Paths.Extended(Path.Combine(dir, "Final files")));
        }
        catch (Exception ex)
        {
            /* If the fixture itself cannot be built, the machine has neither
             * long-path support nor a working \\?\ path, and this test cannot
             * say anything. Skipping silently would be worse than saying so. */
            Assert.Fail($"Could not build the long-path fixture, so this assertion did not " +
                        $"run: {ex.Message}");
            return;
        }

        FixtureSupport.WriteBytes(new byte[] { 1, 2, 3 },
            Paths.Extended(Path.Combine(dir, "Final files", "Final Video.mkv")));
        FixtureSupport.WriteBytes(System.Text.Encoding.UTF8.GetBytes(
            """{ "archive_layout_version": 2, "media_file": "Final files/Final Video.mkv" }"""),
            Paths.Extended(Path.Combine(deep, "manifest.json")));

        Assert.True(dir.Length > 260,
            $"the fixture path is only {dir.Length} characters, so it does not exercise MAX_PATH");

        var index = ArchiveIndex.Scan(_archive);
        var entry = Assert.Single(index.Entries);

        Assert.True(entry.MediaIndex >= 0, "the media file past MAX_PATH was not found");
        Assert.Equal(2, entry.LayoutVersion);
        Assert.Contains(entry.Files, f => f.Rel == "Video metadata/manifest.json");
    }

    // MARK: - Extensions

    [Theory]
    [InlineData("Final files/Final Video.mkv", ".mkv")]
    [InlineData("Final files/Final Video.f137.mp4", ".mp4")]
    [InlineData("Final files/Final Video.MKV", ".mkv")]
    [InlineData("noextension", "")]
    [InlineData(".hidden", "")]
    public void ExtensionUsesTheLastDotAndLowercases(string rel, string expected)
        => Assert.Equal(expected, ArchiveFile.ExtOf(rel));
}
